syncer/dml_worker.go

```diff
@@ -0,0 +1,295 @@
+
+	for i := 0; i < w.workerCount; i++ {
+		causalityJobChs[i] = make(chan *job, w.chanSize)
+		go w.executeCausalityJobs(i, causalityJobChs[i])
+	}
+
+	defer func() {
+		for i := 0; i < w.workerCount; i++ {
+			close(causalityJobChs[i])
+		}
+	}()
+
+	queueBucketMapping := make([]string, w.workerCount)
+	for i := 0; i < w.workerCount; i++ {
+		queueBucketMapping[i] = queueBucketName(i)
+	}
+
+	for j := range w.causalityCh {
+		metrics.QueueSizeGauge.WithLabelValues(w.task, "causality_output", w.source).Set(float64(len(w.causalityCh)))
+		if j.tp == flush || j.tp == conflict {
+			if j.tp == conflict {
+				w.addCountFunc(false, adminQueueName, j.tp, 1, j.targetTable)
+				w.causalityWg.Add(w.workerCount)
+			}
+			// flush for every DML queue
+			for i, causalityJobCh := range causalityJobChs {
+				startTime := time.Now()
+				causalityJobCh <- j
+				metrics.AddJobDurationHistogram.WithLabelValues(j.tp.String(), w.task, queueBucketMapping[i], w.source).Observe(time.Since(startTime).Seconds())
+			}
+			if j.tp == conflict {
+				w.causalityWg.Wait()
+				w.addCountFunc(true, adminQueueName, j.tp, 1, j.targetTable)
+			}
+		} else {
+			queueBucket := int(utils.GenHashKey(j.key)) % w.workerCount
+			w.addCountFunc(false, queueBucketMapping[queueBucket], j.tp, 1, j.targetTable)
+			startTime := time.Now()
+			w.logger.Debug("queue for key", zap.Int("queue", queueBucket), zap.String("key", j.key))
+			causalityJobChs[queueBucket] <- j
+			metrics.AddJobDurationHistogram.WithLabelValues(j.tp.String(), w.task, queueBucketMapping[queueBucket], w.source).Observe(time.Since(startTime).Seconds())
+		}
+	}
+}
+
+// executeCausalityJobs execute jobs in same queueBucket
+// All the jobs received should be executed consecutively.
+func (w *DMLWorker) executeCausalityJobs(queueID int, jobCh chan *job) {
+	jobs := make([]*job, 0, w.batch)
+	workerJobIdx := dmlWorkerJobIdx(queueID)
+	var wg sync.WaitGroup
+	queueBucket := queueBucketName(queueID)
+	for {
+		select {
+		case j, ok := <-jobCh:
+			if !ok {
+				if len(jobs) > 0 {
+					w.logger.Warn("have unexecuted jobs when close job chan!", zap.Any("rest job", jobs))
+				}
+				return
+			}
+			metrics.QueueSizeGauge.WithLabelValues(w.task, queueBucket, w.source).Set(float64(len(jobCh)))
+
+			if j.tp != flush && j.tp != conflict {
+				if len(jobs) == 0 {
+					// set job TS when received first job of this batch.
+					w.lagFunc(j, workerJobIdx)
+				}
+				jobs = append(jobs, j)
+				if len(jobs) < w.batch {
+					continue
+				}
+			}
+
+			// wait for previous jobs executed
+			wg.Wait()
+			batchJobs := jobs
+			wg.Add(1)
+
+			if j.tp == conflict {
+				w.connectionPool.ApplyWithID(w.executeBatchJobs(queueID, batchJobs, func() {
+					wg.Done()
+					w.causalityWg.Done()
+				}))
+			} else {
+				w.connectionPool.ApplyWithID(w.executeBatchJobs(queueID, batchJobs, func() { wg.Done() }))
+			}
+
+			if j.tp == flush {
+				wg.Wait()
+				w.flushCh <- j
+			}
+			jobs = make([]*job, 0, w.batch)
```

Comment:
i am not sure if this will allocate new memory

Reply:
It should allocate new memory here, otherwise, executeBatchJobs may execute wrong jobs.